csrc/id_model/id_graphs.cpp

```diff
@@ -0,0 +1,1768 @@
+// maps.
+IterDomain* IterDomainGraphs::cloneIterDomain(IterDomain* id) {
+  // Figure out which graphs are already initialized to make sure we add the new
+  // expression to them.
+  std::vector<IdMappingMode> initialized_modes;
+  for (auto mode : kIdMappingModes) {
+    auto graph_it = id_graphs_.find(mode);
+    if (graph_it == id_graphs_.end()) {
+      continue;
+    }
+
+    auto& graph = graph_it->second;
+    if (graph.disjointIdSets().disjointSetMap().empty()) {
+      continue;
+    }
+
+    initialized_modes.push_back(mode);
+  }
+
+  auto id_copy = id->cloneWithoutRFactor();
+
+  id_uses_[id_copy] = {};
+  id_definitions_[id_copy] = {};
+
+  for (auto mode : initialized_modes) {
+    idGraph(mode).initializeId(id_copy, {}, {});
+    idGraph(mode).mapIds(id, id_copy);
+  }
+
+  return id_copy;
+}
+
+IdGraph IterDomainGraphs::initializeIdGraph(bool propagate_through_exprs) {
+  IdGraph id_graph(propagate_through_exprs);
+
+  for (auto definition_entry : id_definitions_) {
+    auto id = definition_entry.first;
+    auto defs = definition_entry.second;
+    auto uses_it = id_uses_.find(id);
+    TORCH_INTERNAL_ASSERT(
+        uses_it != id_uses_.end(),
+        "Failed to initialize id: ",
+        id->toString(),
+        " as it's missing a definition entry.");
+    id_graph.initializeId(id, defs, uses_it->second);
+  }
+
+  return id_graph;
+}
+
+void IterDomainGraphs::buildExactMap(const std::vector<Expr*>& exprs) {
+  for (auto expr : exprs) {
+    TensorView* c_tv = ir_utils::getTvOutput(expr);
+
+    auto all_tv_outputs = ir_utils::filterByType<TensorView>(expr->outputs());
+
+    // Map siblings, as all other tv output domains must match the first tv
+    // outputs domain.
+    std::deque<TensorView*> other_tv_outputs(
+        all_tv_outputs.begin(), all_tv_outputs.end());
+    other_tv_outputs.pop_front();
+
+    for (auto other_tv_output : other_tv_outputs) {
+      // Sibling tv's must be exactly mapped with eachother so simply zip
+      // their leaf iter domains.
+
+      TORCH_INTERNAL_ASSERT(
+          other_tv_output->getRootDomain().size() ==
+              c_tv->getRootDomain().size(),
+          "Multiple outputs with mismatched TV domains is not supported.");
+
+      for (auto domain_i : c10::irange(c_tv->getRootDomain().size())) {
+        auto c_id = c_tv->getRootDomain()[domain_i];
+        auto o_id = other_tv_output->getRootDomain()[domain_i];
+        idGraph(IdMappingMode::EXACT).mapIds(o_id, c_id);
+      }
+    }
+
+    // Map producer-consumer relationships based on the root domain map
+    auto tv_inputs = ir_utils::filterByType<TensorView>(expr->inputs());
+    for (auto p_tv : tv_inputs) {
+      // For exact mapings do not map any broadcast dimensions to
+      // non-broadcast dimensions. Prevent any broadcasted axes being mapped
+      // to non-broadcasted axes.
+      auto exact_c2p_root_map =
+          PairwiseRootDomainMap(p_tv, c_tv)
```

Comment:
This is actually not exact. Due to historical reasons, it maps broadcast and non-broadcast domains by default. The true exact mapping requires:

```
PairwiseRootDomainMap(p_tv, c_tv).mapBroadcast(false)
```

That's also what the current ComputeAtMap does.

Due to this broadcast mapping, `computeCoveredGroups` returns some unexpected results. Specifically, since it ignores any group if it has a broadcast domain, even wen the group also may contain non-broadcast domains, some non-broadcast domains are missing in the covered groups, which means that `loop_graph_copy_promotion_map` may not be accurate.

Interestingly, test `Indexing19` does seem to work as intended even with this "exact" map, and adding `mapBroadcast(false)` actually results in unexpected loop promotion mappings.

I think this is a bug. Will work on it.